aggregator/election_mgr.go

```diff
@@ -0,0 +1,518 @@
+		campaignOpts:    campaignOpts,
+		electionOpts:    opts.ElectionOptions(),
+		campaignRetrier: campaignRetrier,
+		changeRetrier:   changeRetrier,
+		electionKeyFmt:  opts.ElectionKeyFmt(),
+		leaderService:   opts.LeaderService(),
+		leaderValue:     campaignOpts.LeaderValue(),
+		state:           electionManagerNotOpen,
+		doneCh:          make(chan struct{}),
+		notifyCh:        make(chan struct{}, 1),
+		electionState:   FollowerState,
+		sleepFn:         time.Sleep,
+		metrics:         newElectionManagerMetrics(instrumentOpts.MetricsScope()),
+	}
+	go mgr.reportMetrics()
+	return mgr
+}
+
+func (mgr *electionManager) Open(shardSetID string) error {
+	mgr.Lock()
+	defer mgr.Unlock()
+
+	if mgr.state != electionManagerNotOpen {
+		return errElectionManagerExpectNotOpen
+	}
+	mgr.state = electionManagerOpen
+	mgr.electionKey = fmt.Sprintf(mgr.electionKeyFmt, shardSetID)
+	go mgr.startElectionLoop()
+	return nil
+}
+
+func (mgr *electionManager) ElectionState() ElectionState {
+	mgr.RLock()
+	electionState := mgr.electionState
+	mgr.RUnlock()
+	return electionState
+}
+
+func (mgr *electionManager) Resign(ctx context.Context) error {
+	mgr.Lock()
+	if mgr.state != electionManagerOpen {
+		mgr.Unlock()
+		return errElectionManagerExpectOpen
+	}
+	if mgr.electionState == FollowerState {
+		mgr.Unlock()
+		mgr.metrics.followerResign.Inc(1)
+		return nil
+	}
+	mgr.state = electionManagerResigning
+	mgr.resignWg = &sync.WaitGroup{}
+	resignWg := mgr.resignWg
+	resignWg.Add(1)
+	mgr.Unlock()
+
+	err := mgr.leaderService.Resign(mgr.electionKey)
+	if err != nil {
+		resignWg.Done()
+		mgr.Lock()
+		mgr.state = electionManagerOpen
+		mgr.resignWg = nil
+		mgr.Unlock()
+		err := fmt.Errorf("leader service resign failed: %v", err)
+		mgr.metrics.resignErrors.Inc(1)
+		mgr.logError("resign failed", err)
+		return err
+	}
+
+	// Wait for TTL to give the follower instance some time to take over the
+	// leader role before starting a new campaign after resigning.
+	mgr.sleepFn(time.Duration(mgr.electionOpts.TTLSecs()) * time.Second)
+	resignWg.Done()
+
+	// Now wait for the current instance to become a follower.
+	var (
+		ctxDone    = false
+		isFollower = false
+	)
+
+	for {
+		select {
+		case <-ctx.Done():
+			ctxDone = true
+		case <-mgr.notifyCh:
+			isFollower = mgr.ElectionState() == FollowerState
+		}
+		if ctxDone || isFollower {
+			break
+		}
+	}
+
+	// If the context expires, we cancel the in progress change and re-check
+	// the election state.
+	if !isFollower {
+		mgr.cancelInProgressChange()
+		isFollower = mgr.ElectionState() == FollowerState
+	}
+
+	mgr.Lock()
+	mgr.state = electionManagerOpen
+	mgr.resignWg = nil
+	mgr.Unlock()
+
+	if isFollower {
+		return nil
+	}
+
+	// If the election state is not follower, it means we failed to establish
+	// the follower state with kv.
+	err = fmt.Errorf("instance resigned but is not follower: %v", ctx.Err())
+	mgr.metrics.resignNotFollower.Inc(1)
+	mgr.logError("resign failed", err)
+	return err
```

Comment:
At this point will the process still believe its a leader?  Seems strange and wondering what the correct course of action the caller can make at this point.